content/blog/2022-04-07-compiling-black-with-mypyc-part-1.md

```diff
@@ -0,0 +1,591 @@
+
+- Types are enforced at runtime and violations **will** cause a TypeError (this makes
+  using `typing.Any` quite dangerous).
+
+- Compiled modules can't be run directly, you have to import 'em instead.
+
+- Monkeypatching anything compiled is unlikely to work as compiled code skips many of the
+  runtime `__dict__` lookups normal Python does.
+
+- Assignments to class and instance namespaces will either error or do nothing, in
+  particular you can't add previously undeclared attributes later on.
+
+- Standard multiple inheritance isn't supported.
+
+- Profilers (e.g. CProfile), debuggers (e.g. pdb), and tracing hooks (e.g. coverage) won't
+  work as the C code doesn't trigger the relevant hooks.
+
+There's more details available in the official
+[mypyc documentation here][mypyc-deviations-one] and [also here][mypyc-deviations-two].
+
+## black + mypyc: Initial steps
+
+I wasn't the first person to integrate mypyc into Black, way back in September 2019,
+[@msullivan] [opened a PR getting Black ready][original-pr] for compilation. Unfortunately
+as typical in open source projects, no one took the half-completed work and pushed it to
+production, i.e. PyPI, ... for almost two years.
+
+> You might be wondering why performance even matters, well clearly it mattered a lot
+> since [GH-366] was opened in June 2018! The TL;DR is that for environments where Black
+> is **run on save automatically, the more responsive it is the better** as less time
+> is spent in a laggy editor window.
+>
+> Start up time is important too given imports are costly, but this issue was explicitly
+> about formatting throughput (it turns out mypyc can reduce import time too!).
+
+Given I first publicly announced my work finishing up the project on July 4th 2021, I
+sadly don't remember why I decided to pick it up. All I remember was fighting mypyc
+compile-time type errors and fighting an outdated gcc... well OK I do remember I was
+looking to learn more about type systems and C development in general, but that's it, I
+promise!
+
+Anyway, even upgrading from gcc-9 to gcc-10 wasn't enough. Currently mypyc has a bug
+([mypyc#885]) where its boxing / unboxing code triggers
+`array subscript 1 is above array bounds` on
+[`blib2to3.pgen2.parse.Parser.setup`][gcc-bug-code]. I managed to simplify the
+reproduction case to the following:
+
+```python3
+from typing import Optional, Tuple
+
+Context = Tuple[str, Tuple[int, int]]
+RawNode = Tuple[int, Optional[str], Optional[Context]]
+
+def setup() -> None:
+    # This is what ticks off gcc.
+    newnode: RawNode = (7, None, None)
+```
+
+... yeah I know this doesn't make mypyc seem particularly robust. While I would argue it's
+not entirely stupid to use it in prod (because I did so), you would be right in pointing
+out it's alpha quality and needs careful testing. And to be fair, poor past me didn't know
+just how many issues I'd soon face. All I can hope is that this work will help mypyc
+improve ❀
```

Comment:
```suggestion
improve ❀.
```
Maybe? Not sure how it flows with the flower.

Reply:
Yea, I think I can get away with not using a period here :)